wethegit/corgi
# Add support for deep-merge of YAML data

## Overview
This adds support for **nested** locale-specific overrides of YAML data. 
Closes #104 

Also included:
- minor dependabot fixes which are already in `main`
- minor changes to variable naming in `page-setup.js`
- minor updates to documentation 

## What problem does this solve?
Currently, overriding YAML data from the `index.yml` file simply overwrites the first identical key. For example, if you had:
```yml
# index.yml
section:
  title: Our clients love us
  body: Lorem ipsum dolor…
  clientList: ['amazon', 'canadian tire', 'best buy']
```
and you wanted to alter the client list for the UK version of the section:
```yml
# en-gb.yml
section:
  clientList: ['smyths', 'game']
```
You would lose the `title` and `body` properties entirely. The `section` key would only contain your new `clientList`.

## Solution
Continuing with the example above, this update will *preserve* those `title` and `body` key/value pairs. This works with deeply-nested objects as well.

This has many implications for speeding up workflow around localization. Exciting stuff.

## Example
The video below shows an example of this in action on a fresh corgi project. I'm defining all the "global" navigation items for an example app, and then only overwriting the `label` property of each, for the locale-specific page render. The idea here is that the URL enpoints are always the same, so there's no need to re-declare them on every locale file.

https://github.com/wethegit/corgi/assets/30575213/eda04117-428f-4aeb-acba-13b78d1b2273




(shout out to @liamegan for tag-teaming this deepMerge function with me!)

## Testing this
The easiest way to test this update is to reference the package binary directly, and bootstrap a sample project:
1. Check out this branch by running `git checkout feature/template/yaml-deep-merge`
2. `cd` into the project's root directory if you're not already there, and run `node ./cli/bin/index.js project test-corgi-build`
3. Run the following commands to start the project: 
```shell
cd test-corgi-build
nvm use # or n auto
npm install
npm start
```
